### PR TITLE
Artifact Rejection

### PR DESCRIPTION
**Pull request by orbitalhybridization**
### Artifact Rejection
_________________________________________________________________________
This PR adds a new subfolder to the signal module: evaluate. The subfolder contains two classes, Evaluator and Rules, which are the pre-task handlers for Artifact Rejection. There were also some small additions to the parameters.json file, and a draft test file for evaluator.

### Updated
_________________________________________________________________________

- signal module has new subfolder, evaluate
- Evaluator (Class)
      - maintains rules enabled and tests raw data against rules.
      - has an 'evaluate' method that simply returns False upon rule breakage.
- Rules (Class)
       - a Python Abstract Base Class (ABC) that currently has two subclasses: HighVoltage and 
       LowVoltage
       - Utilises abstract method, 'is_broken,' for all rules, which contains instructions on how the rule is 
        evaluated
- Changes to parameters.json, adding toggle options for Artifact Rejection, HighVoltage Threshold, and LowVoltage Threshold, as well as integer entry fields for voltage thresholds if enabled.

### CODE REVIEW

File: bcipy/signal/model/offline_analysis.py [hunk 1]
```diff
@@ -84,7 +84,9 @@ def offline_analysis(data_folder: str = None,
                                 channel_map=channel_map,
                                 trial_length=trial_length)
 
+    # train model
```
Comment (tab-cmd) [inline bcipy/signal/model/offline_analysis.py]: remove this!

File: bcipy/signal/tests/evaluate/test_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.channels = 32
+
+    def tearDown(self):
+        unstub()
+
+    def test_init_rules(self):
+        """Test init of ruleset. We expect that rules enabled in the
+        params are initialized as part of the ruleset"""
+
+        if self.expected_high_voltage:
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_evaluator.py:30]: remove if statements and make separate tests

File: bcipy/signal/tests/evaluate/test_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            self.assertIsInstance(self.evaluator.rules[0], HighVoltage)
+
+        if self.expected_low_voltage:
+            self.assertIsInstance(self.evaluator.rules[1], LowVoltage)
+
+    def test_evaluate_true(self):
+        """Test evaluate with is_broken returning True"""
+        when(self.evaluator.rules[0]).is_broken(any()).thenReturn(True)
+        when(self.evaluator.rules[1]).is_broken(any()).thenReturn(True)
+        data = gen_random_data(-1,2,self.channels)
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_evaluator.py]: linting!

File: bcipy/signal/tests/evaluate/test_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        self.assertFalse(self.evaluator.evaluate(data))
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_evaluator.py:44]: I'd rename this test and comment on expected behaviour

File: bcipy/signal/tests/evaluate/test_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    def test_evaluate_false(self):
+        """Test evaluate with is_broken returning False """
+        when(self.evaluator.rules[0]).is_broken(any()).thenReturn(False)
+        when(self.evaluator.rules[1]).is_broken(any()).thenReturn(False)
+        data = gen_random_data(-1,2,self.channels)
+        self.assertIsNone(self.evaluator.evaluate(data))
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_evaluator.py]: why none? aren't we returning True/False?

File: bcipy/signal/tests/evaluate/test_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    def test_evaluate_true(self):
+        """Test evaluate with is_broken returning True"""
+        when(self.evaluator.rules[0]).is_broken(any()).thenReturn(True)
+        when(self.evaluator.rules[1]).is_broken(any()).thenReturn(True)
+        data = gen_random_data(-1,2,self.channels)
+        self.assertFalse(self.evaluator.evaluate(data))
+
+    def test_evaluate_false(self):
+        """Test evaluate with is_broken returning False """
+        when(self.evaluator.rules[0]).is_broken(any()).thenReturn(False)
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_evaluator.py:49]: Great use of mocking!

File: bcipy/signal/tests/evaluate/test_rules.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        data = gen_random_data(-0.5,0.5,self.channels)
+        self.assertFalse(self.lowvoltage_rule.is_broken(data))
+
+    def test_high_voltage_failing_signal(self):
+        """Test generated signal with one data point above threshold """
+        data = gen_random_data(-5,0,self.channels)
+        #ascertain that at least one random datapoint is above threshold to test np.amax edgecase
+        data[np.random.randint(self.channels)] = 1.5
+        self.assertTrue(self.highvoltage_rule.is_broken(data))
+
```
Comment (tab-cmd) [inline bcipy/signal/tests/evaluate/test_rules.py:54]: 👍

**Review by tab-cmd — CHANGES_REQUESTED** [review]
looking good- almost ready to merge. a few comments and you should run linting!

File: bcipy/signal/model/offline_analysis.py (hunk 1 above)
Comment (orbitalhybridization) [reply bcipy/signal/model/offline_analysis.py]: Gotcha!

File: bcipy/signal/tests/evaluate/test_evaluator.py (hunk 2 above)
Comment (orbitalhybridization) [reply bcipy/signal/tests/evaluate/test_evaluator.py:30]: that makes sense, I've separated them into two different tests!

File: bcipy/signal/tests/evaluate/test_evaluator.py (hunk 3 above)
Comment (orbitalhybridization) [reply bcipy/signal/tests/evaluate/test_evaluator.py]: got it! flake8 flagged the missing whitespace after ',' error there.

File: bcipy/signal/tests/evaluate/test_evaluator.py (hunk 5 above)
Comment (orbitalhybridization) [reply bcipy/signal/tests/evaluate/test_evaluator.py]: Originally I didn't have evaluate default to anything, but it is more intuitive to default to True, so I've changed the method to fit that! Thank you :D

File: bcipy/signal/tests/evaluate/test_evaluator.py (hunk 4 above)
Comment (orbitalhybridization) [reply bcipy/signal/tests/evaluate/test_evaluator.py:44]: Ah, yeah it's a little confusing especially considering the behaviour you mention in the next comment. Let me know what you think of the changes!

**Comment by smektala** [comment]
Looks good to me!

**Review by smektala — APPROVED** [review]

**Review by tab-cmd — APPROVED** [review]

**Pull request merged**